sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -488,90 +493,70 @@
-                                  size),
-                              timestamp,
-                              currentWindow,
-                              currentElement.getPane()));
+                public double getProgress() {
+                  return splitDelegate.getProgress();
                 }
               };
-        } else {
-          mainInputConsumer = this::processElementForSplitRestriction;
-          // OutputT == RestrictionT
           this.processContext =
-              new NonWindowObservingProcessBundleContext() {
+              new SizedRestrictionWindowObservingProcessBundleContext(
+                  PTransformTranslation.SPLITTABLE_TRUNCATE_SIZED_RESTRICTION_URN);
+        } else {
+          mainInputConsumer =
+              new SplittableFnDataReceiver() {
+                private final HandlesSplits splitDelegate =
+                    (HandlesSplits) Iterables.get(mainOutputConsumers, 0);
+
+                @Override
+                public void accept(WindowedValue input) throws Exception {
+                  processElementForTruncateRestriction(input);
+                }
+
                 @Override
-                public void outputWithTimestamp(OutputT output, Instant timestamp) {
-                  double size =
-                      doFnInvoker.invokeGetSize(
-                          new DelegatingArgumentProvider<InputT, OutputT>(
-                              this,
-                              PTransformTranslation.SPLITTABLE_SPLIT_AND_SIZE_RESTRICTIONS_URN
-                                  + "/GetSize") {
-                            @Override
-                            public Object restriction() {
-                              return output;
-                            }
-
-                            @Override
-                            public Instant timestamp(DoFn<InputT, OutputT> doFn) {
-                              return timestamp;
-                            }
-
-                            @Override
-                            public RestrictionTracker<?, ?> restrictionTracker() {
-                              return doFnInvoker.invokeNewTracker(this);
-                            }
-                          });
-
-                  outputTo(
-                      mainOutputConsumers,
-                      (WindowedValue<OutputT>)
-                          WindowedValue.of(
-                              KV.of(
-                                  KV.of(
-                                      currentElement.getValue(),
-                                      KV.of(output, currentWatermarkEstimatorState)),
-                                  size),
-                              timestamp,
-                              currentElement.getWindows(),
-                              currentElement.getPane()));
+                public SplitResult trySplit(double fractionOfRemainder) {
+                  return splitDelegate.trySplit(fractionOfRemainder);
```

Comment:
We should return null here and leave a todo BEAM-10303 for window observing optimization.

Because of window exploding, we'll only split on the current window which will return the wrong primary/residual since it won't take into account these other windows that have been fully processed or yet to be started.